src/blog/understanding-ecmascript-part-3.md

```diff
@@ -0,0 +1,362 @@
+
+> [`AsyncFunctionBody :`](https://tc39.es/ecma262/#prod-AsyncFunctionBody)
+> <code>FunctionBody<sub>[~Yield, +Await]</sub></code>
+
+Note that `AsyncFunctionBody` has no parameters &mdash; they get added to the `FunctionBody` on the right-hand side.
+
+If we expand this production, we get:
+
+> `AsyncFunctionBody :`
+> `FunctionBody_Await`
+
+In other words, async functions have `FunctionBody_Await`, meaning a function body where `await` is treated as a keyword.
+
+On the other hand, if we're inside a non-async function, the relevant production is:
+
+> [<code>FunctionDeclaration<sub>[Yield, Await, Default]</sub>](https://tc39.es/ecma262/#prod-FunctionDeclaration) :</code>
+> <code>function BindingIdentifier<sub>[?Yield, ?Await]</sub> ( FormalParameters<sub>[~Yield, ~Await]</sub> ) { FunctionBody<sub>[~Yield, ~Await]</sub> }</code>
+
+(`FunctionDeclaration` has another production, but it's not relevant for our code example.)
+
+To avoid combinatorial expansion, let's ignore the `Default` parameter which is not used in this particular production.
+
+The expanded form of the production is:
+
+> `FunctionDeclaration :`
+> `function BindingIdentifier ( FormalParameters ) { FunctionBody }`
+
+> `FunctionDeclaration_Yield :`
+> `function BindingIdentifier_Yield ( FormalParameters_Yield ) { FunctionBody }`
+
+> `FunctionDeclaration_Await :`
+> `function BindingIdentifier_Await ( FormalParameters_Await ) { FunctionBody }`
+
+> `FunctionDeclaration_Yield_Await :`
+> `function BindingIdentifier_Yield_Await ( FormalParameters_Yield_Await ) { FunctionBody }`
+
+In this production we always get `FunctionBody` (without `_Yield` and without `_Await`), since the `FunctionBody` in the non-expanded production is parameterized with `[~Yield, ~Await]`.
+
+Function name and formal parameters are treated differently: they get the parameters `_Await` and `_Yield` if the left-hand side symbol has them.
+
+To summarize: Async functions have a `FunctionBody_Await` and non-async functions have a `FunctionBody` (without `_Await`). Since we're talking about non-generator functions, both our async example function and our non-async example function are parameterized without `_Yield`.
+
+Maybe it's hard to remember which one is `FunctionBody` and which `FunctionBody_Await`. Is `FunctionBody_Await` for a function where `await` is an identifier, or for a function where `await` is a keyword?
+
+You can think of the `_Await` parameter meaning "`await` is a keyword". This approach is also future proof. Imagine a new keyword, `blob` being added, but only inside "blobby" functions. Non-blobby non-async non-generators would still have `FunctionBody` (without `_Await`, `_Yield` or `_Blob`), exactly like they have now. Blobby functions would have a `FunctionBody_Blob`, async blobby functions would have `FunctionBody_Await_Blob` and so on.
```

Comment:
Well, future proof up to still having to add ~Blob to FunctionBody, I guess?

Reply:
Right, I meant that the expanded forms won't change.

Reply:
Added: We'd still need to add the Blob subscript to the rules, but the expanded forms of FunctionBody for already existing functions stay the same.